### PR TITLE
actix-cors: Create `allowed_origin_fn` builder method

### PR DESCRIPTION
**Pull request by pldin601**
## PR Type

Feature

## PR Checklist
Check your PR fulfills the following:

- [x] Tests for the changes have been added / updated.
- [x] Documentation comments have been added / updated.
- [x] A changelog entry has been made for the appropriate packages.
- [x] Format code with the latest stable rustfmt

## Overview

Inspired by: https://www.npmjs.com/package/cors#configuring-cors-w-dynamic-origin

This PR adds support of dynamic origin. A predicate functions can be used to dynamically control `Access-Control-Allow-Origin` response header depending on business logic.

It will be called on each request if client's `Origin` request header wasn't covered by any existing `allowed_origin` rules.
If function returned `true`, client's `Origin` request header will be echoed back in the `Access-Control-Allow-Origin` response header, otherwise no additional action will be performed.

### CODE REVIEW

File: actix-cors/CHANGES.md [hunk 1]
```diff
@@ -7,6 +7,7 @@
 * Update `actix-web` dependency to 3.0.0.
 * Minimum supported Rust version (MSRV) is now 1.42.0.
 * Implement the Debug trait on all public types.
+* Implement `allowed_origin_fn` builder method.
```
Comment (robjtede) [inline actix-cors/CHANGES.md:10]: move to unreleased section

File: actix-cors/src/lib.rs [hunk 2]
```diff
@@ -50,6 +50,8 @@ use std::iter::FromIterator;
 use std::rc::Rc;
 use std::task::{Context, Poll};
 
+use core::fmt;
```
Comment (robjtede) [inline actix-cors/src/lib.rs:53]: use std::fmt

Comment (robjtede) [inline actix-cors/src/lib.rs]:
```suggestion

use std::fmt;
```

File: actix-cors/src/lib.rs [hunk 3]
```diff
@@ -602,6 +636,10 @@ impl Inner {
         }
     }
 
+    fn validate_origin_fn(&self, req: &RequestHead) -> bool {
```
Comment (robjtede) [inline actix-cors/src/lib.rs:639]:
```suggestion
    fn validate_origin_fns(&self, req: &RequestHead) -> bool {
```

File: actix-cors/src/lib.rs [hunk 4]
```diff
@@ -1205,4 +1245,66 @@ mod tests {
...
+            resp.headers()
+                .get(header::ACCESS_CONTROL_ALLOW_ORIGIN)
+                .unwrap()
+                .to_str()
+                .unwrap()
+        );
+    }
+
+    #[actix_rt::test]
+    async fn test_not_allowed_origin_fn() {
```
Comment (robjtede) [inline actix-cors/src/lib.rs:1283]: This test should include a successful call from example.com origin.

File: actix-cors/src/lib.rs [hunk 5]
```diff
@@ -1205,4 +1245,66 @@ mod tests {
                 .as_bytes()
         );
     }
+
+    #[actix_rt::test]
+    async fn test_allowed_origin_fn() {
```
Comment (robjtede) [inline actix-cors/src/lib.rs:1250]: This test should include a successful call from example.com origin.

File: actix-cors/src/lib.rs [hunk 6]
```diff
@@ -268,6 +272,18 @@ impl Cors {
         self
     }
 
+    /// Add an predicate function that will be called on each request if client's `Origin`
```
Comment (robjtede) [inline actix-cors/src/lib.rs:275]:
```suggestion
    /// Add a predicate function that will be called on each request if client's `Origin`
```

File: actix-cors/src/lib.rs [hunk 7]
```diff
@@ -50,6 +50,8 @@ use std::iter::FromIterator;
 use std::rc::Rc;
 use std::task::{Context, Poll};
```
Comment (robjtede) [inline actix-cors/src/lib.rs:52]: nit:

```suggestion
```

File: actix-cors/src/lib.rs [hunk 8]
```diff
@@ -567,10 +583,21 @@ pub struct CorsMiddleware<S> {
     inner: Rc<Inner>,
 }
 
+struct OriginFn {
+    f: fn(req: &RequestHead) -> bool,
+}
+
+impl fmt::Debug for OriginFn {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        return write!(f, "origin_fn");
```
Comment (robjtede) [inline actix-cors/src/lib.rs:592]:
```suggestion
        write!(f, "origin_fn")
```

**Review by robjtede — CHANGES_REQUESTED** [review]
Also think it'd be good to include some info in the main docs on Cors too. It should be crystal clear to readers that the _fns are called AFTER allowed origin whitelist checks when none have matched.

**Comment by pldin601** [comment]
> Also think it'd be good to include some info in the main docs on Cors too. It should be crystal clear to readers that the _fns are called AFTER allowed origin whitelist checks when none have matched.

Can you suggest me where I can add this documentation exactly?

**Comment by robjtede** [comment]
> Can you suggest me where I can add this documentation exactly?

On the Cors struct docs, such that it appears on this page at the top: https://docs.rs/actix-cors/0.3.0/actix_cors/struct.Cors.html

**Review by Neopallium — APPROVED** [review]
looks good.

**Review by robjtede — APPROVED** [review]

**Comment by robjtede** [comment]
Nice feature, thanks for this 👍

**Pull request merged**